sw-emulator/lib/types/src/axi_bus.rs

```diff
@@ -0,0 +1,62 @@
+/*++
+
+Licensed under the Apache-2.0 license.
+
+File Name:
+
+    axi_bus.rs
+
+Abstract:
+
+    File contains definition of the AxiBus trait.
+
+--*/
+
+use crate::{AxiAddr, AxiData, AxiSize};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum AxiBusError {
+    /// Instruction access exception
+    InstrAccessFault,
+
+    /// Load address misaligned exception
+    LoadAddrMisaligned,
+
+    /// Load access fault exception
+    LoadAccessFault,
+
+    /// Store address misaligned exception
+    StoreAddrMisaligned,
+
+    /// Store access fault exception
+    StoreAccessFault,
+}
+
+/// Represents an abstract AXI memory bus. Used to read and write from RAM and
+/// peripheral addresses.
+pub trait AxiBus {
```

Comment:
I absolutely hate this copy-paste. Why didn't we just increase the address size to 64-bit everywhere?

Reply:
(assuming the only reason for this change is the increased word size): my preference:

1. Use 64-bit addresses and data in Bus trait everywhere; the 32-bit parts could ignore the high bits and error when size is 64-bit.
2. If that's not workable for some reason we could add associated types to the existing bus trait:

```rs
trait Bus {
  type Addr: num_traits::Num + num_traits::Unsigned + From<u32>;
  type Data: num_traits::Num + num_traits::Unsigned + From<u32>;

  fn read(&mut self, addr: Self::Addr) -> Result<Self::Data, ()>;
  // ...
}
```

Reply:
poll and reset() could be moved to their own separate trait if necessary.

Reply:
I agree I don't love the copy-pasta, especially for the derive macros and the register types.

I like the idea of 64-bit everywhere, maybe either with a type parameter or a function that specifies whether 64-bit is allowed.

Another idea that I think would be okay: keep the separate `AxiBus` trait for read64/write64, but don't support derive. Instead, use autogeneration to automatically create the `AxiBus` implementation. (This is the approach we are mostly using in the MCU, for example: https://github.com/chipsalliance/caliptra-mcu-sw/blob/main/registers/generated-emulator/src/root_bus.rs).

Or only support the dynamic bus.

Since the AXI bus doesn't need to be quite as flexible as the main bus, I think it might be okay to go the autogenerated route.

Reply:
@korran both these approaches require a lot of code changes all across the emulator code.

I think the simplest things is to implement the toplevel axi_root_bus by hand, which dispatches to Bus.